### PR TITLE
feat: Initial commit for Memory manager

### PR DESCRIPTION
**Pull request by jona62**
## Description

Added the Memory manager functionality and unit tests, including tests for memory creation, strategy management, error handling, and the Memory class wrapper. The test suite covers all major methods and edge cases with proper mocking of AWS clients.

## Type of Change

• [x] New feature (non-breaking change which adds functionality)
• [ ] Bug fix (non-breaking change which fixes an issue)
• [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
• [ ] Documentation update
• [ ] Performance improvement
• [ ] Code refactoring

## Testing

• [x] Unit tests pass locally
• [ ] Integration tests pass (if applicable)
• [x] Test coverage remains above 80%
• [x] Manual testing completed

## Checklist

• [x] My code follows the project's style guidelines (ruff/pre-commit)
• [x] I have performed a self-review of my own code
• [x] I have commented my code, particularly in hard-to-understand areas
• [ ] I have made corresponding changes to the documentation
• [x] My changes generate no new warnings
• [x] I have added tests that prove my fix is effective or that my feature works
• [x] New and existing unit tests pass locally with my changes
• [ ] Any dependent changes have been merged and published

## Security Checklist

• [x] No hardcoded secrets or credentials
• [x] No new security warnings from bandit
• [x] Dependencies are from trusted sources
• [x] No sensitive data logged

## Breaking Changes

N/A

## Additional Notes

This test suite provides comprehensive coverage for the Memory manager including:
• Memory creation and lifecycle management
• Strategy operations (add, modify, delete)
• Error handling for various AWS client exceptions
• Memory class wrapper functionality
• Proper mocking of boto3 clients and responses
• Edge cases and validation scenarios

### CODE REVIEW

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py [hunk 1]
```diff
@@ -0,0 +1,986 @@
...
+
+    def _create_memory(
+        self,
+        name: str,
+        strategies: Optional[List[Dict[str, Any]]] = None,
+        description: Optional[str] = None,
+        event_expiry_days: int = 90,
+        memory_execution_role_arn: Optional[str] = None,
+    ) -> 'Memory':
+        """Creaete a memory resource and return the raw response.
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: typo

File: src/bedrock_agentcore_starter_toolkit/operations/memory/constants.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+"""Constants for Bedrock AgentCore Memory SDK."""
+
+from enum import Enum
+from typing import Dict, List
+
+
+class StrategyType(Enum):
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/constants.py:7]: I would avoid multiple enums and wrapper keys by defining method for with same enum.

1. def extraction_wrapper_key(self) -> str:
2. def consolidation_wrapper_key(self) -> str:
3. def get_memory_strategy(self)->str:
4. def get_override_type(self)->str:

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py [hunk 3]
```diff
@@ -0,0 +1,986 @@
...
+                else:
+                    logger.debug("Memory status: %s (%d seconds elapsed)", status, elapsed)
+
+            except ClientError as e:
+                logger.error("Error checking memory status: %s", e)
+                raise
+
+            time.sleep(poll_interval)
+
+        raise TimeoutError("Memory %s did not become ACTIVE within %d seconds" % (memory_id, max_wait))
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: use f-strings for better readability.

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py [hunk 4]
```diff
@@ -0,0 +1,986 @@
...
+        return self._memory[key]
+
+    def _get(self, key: str, default=None) -> Any:
+        """Provides dict.get() style access to memory fields"""
+        return self._memory.get(key, default)
+
+    def __repr__(self):
+        return self._memory.__repr__()
+
+    def delete(self):
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: Make it side-affect prone where underlying object is being mutated. I would stick to pattern of memory manager operating on provided memory object. Memory should be a pure data object.

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py [hunk 5]
```diff
@@ -0,0 +1,986 @@
...
+        }
+
+        if description:
+            strategy[StrategyType.SEMANTIC.value]["description"] = description
+        if namespaces:
+            strategy[StrategyType.SEMANTIC.value]["namespaces"] = namespaces
+
+        return self._add_strategy(memory_id, strategy)
+
+    def add_semantic_strategy_and_wait(
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py:475]: instead of two methods why not just have add_semantic_strategy with polling as optional param?

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py [hunk 6]
```diff
@@ -0,0 +1,986 @@
...
+        namespaces = strategy_config.get("namespaces", [])
+        for namespace in namespaces:
+            self._validate_namespace(namespace)
+
+class Memory:
+    """
+    Represents a single, provisioned AgentCore Memory resource. 
+    This class provides convenient delegation to MemoryManager operations.
+    """
+    def __init__(self, memory: Dict[str, Any], manager: MemoryManager):
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: wouldn't memory builder be a better pattern here - memory = (MemoryBuilder()
         .name("test")
         .with_semantic_strategy("strategy1")
         .build())

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/MemoryStrategy.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+from typing import Any, Dict
+
+class MemoryStrategy:
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/models/MemoryStrategy.py]: I would just define use dataclass/Pydantic model. How does a developer know fields?

**Review by Vivekbhadauria1 — CHANGES_REQUESTED** [review]

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py (hunk 3 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: Fixed

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py (hunk 5 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py:475]: This was just a design decision I carried over from the `MemoryClient` and I think optional parameters are often ignored but making an explicit function adds more visibility for customers to discover this function. Let me know your thoughts, I'm indifferent about either

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py (hunk 4 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: I had the same thoughts as well I figured to follow the same pattern as other python SDKs where the created resource object has the ability to act on itself. But Given the manager has the same functionality I can see how it becomes problematic because a user can use the manager to change the memory but it's no longer the same resource in the memory object.

For instance:

```python
from google.cloud import storage
storage_client = storage.Client()

bucket = storage_client.get_bucket(bucket_name)
blob = bucket.blob("my-file.txt")
blob.upload_from_string("Hello, Cloud Storage!")
print(f"File my-file.txt uploaded to {bucket_name}.")
```

File: src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py (hunk 6 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/manager.py]: Based on the earlier comment, the builder pattern would be handy if we're going to continue using Memory as a stateful class. But since it'll be removed, Memory will strictly be a data class which has the response of a CREATE and UPDATE memory API invocations.

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/MemoryStrategy.py (hunk 7 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/models/MemoryStrategy.py]: I recently discovered you could use this to add autocompletion for fields in a class, but I'll try to see if Pydantic would solve this without having to explicitly define the response variables(Trying to avoid translating the response fields)
```python
def __dir__(self):
    """Enable tab completion and introspection of available attributes"""
    return list(self._data.keys()) + ['get']
```

File: tests_integ/memory/memory-manager.ipynb [hunk 8]
```diff
@@ -0,0 +1,332 @@
+{
```
Comment (Vivekbhadauria1) [inline tests_integ/memory/memory-manager.ipynb:1]: Please move the file to "tests_integ"

File: src/bedrock_agentcore_starter_toolkit/operations/memory/constants.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        consolidation_keys = {
+            OverrideType.SEMANTIC_OVERRIDE: "semanticConsolidationOverride",
+            OverrideType.SUMMARY_OVERRIDE: "summaryConsolidationOverride",
+            OverrideType.USER_PREFERENCE_OVERRIDE: "userPreferenceConsolidationOverride",
+        }
+        return consolidation_keys.get(self)
+
+
+# Legacy enum for backward compatibility - can be removed in future versions
+class MemoryStrategyTypeEnum(Enum):
```
Comment (Vivekbhadauria1) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/constants.py]: This is a new class, why do we need legacy support?

**Review by Vivekbhadauria1 — CHANGES_REQUESTED** [review]

File: tests_integ/memory/memory-manager.ipynb (hunk 8 above)
Comment (jona62) [reply tests_integ/memory/memory-manager.ipynb:1]: Done

File: src/bedrock_agentcore_starter_toolkit/operations/memory/constants.py (hunk 9 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/constants.py]: Left over artifact from testing
